### PR TITLE
use pow implementation with incremental multiplicative base

### PR DESCRIPTION
**Pull request by vhvb1989**

### CODE REVIEW

File: sdk/core/core/src/az_json_parser.c [hunk 1]
```diff
@@ -131,33 +131,39 @@ typedef struct
...
+}
+
 // try to convert a number to double
 // Return OVERFLOW when exp goes higher than 19 (double can hold 2 ^ 64 / 2, so 10 ^ 19 would be the
 // max exp supported)
 AZ_NODISCARD az_result az_json_number_to_double(az_dec_number const* p, double* out)
 {
   if (p->exp >= 19)
   {
+    // Can't hold more than 10^18 into double
```
Comment (gilbertw) [inline sdk/core/core/src/az_json_parser.c]: This comment and the function comment disagree

File: sdk/core/core/src/az_json_parser.c [hunk 2]
```diff
@@ -131,33 +131,39 @@ typedef struct
...
+// calculate 10 ^ exp with O(log exp) by doing incremental multiplication
+AZ_NODISCARD static double _ten_to_exp(int16_t exp)
+{
+  double result = 1;
+  double incrementing_base = 10;
+  int16_t abs_exp = exp < 0 ? -exp : exp;
+
+  while (abs_exp > 0)
+  {
+    // non even exp would update result to current incremented base
```
Comment (gilbertw) [inline sdk/core/core/src/az_json_parser.c]: odd instead of non-even

File: sdk/core/core/test/main.c [hunk 3]
```diff
@@ -330,6 +330,12 @@ int main()
     az_json_token token;
     TEST_ASSERT(az_json_parser_parse_token(&state, &token) == AZ_ERROR_BUFFER_OVERFLOW);
   }
+  {
+    az_json_parser state = { 0 };
+    TEST_EXPECT_SUCCESS(az_json_parser_init(&state, AZ_SPAN_FROM_STR("10000000000000000000000e17")));
```
Comment (gilbertw) [inline sdk/core/core/test/main.c]: Please beef up these tests before checking in.  pos/neg exp, boundary conditions, exp == 0, 1, -1, ....

Comment (vhvb1989) [reply sdk/core/core/test/main.c]: we already had test for that.
But, I am taking the chance to refactor tests files for easier understanding

**Comment by vhvb1989** [comment]
@gilbertw , C99 used IEEE_754 for handling floating point
https://en.wikipedia.org/wiki/IEEE_754
So, We don't really need to check for overflow or underflow, since C99 would return `inf` or 0..
I have added tests for that... so, we just stick to that behavior
